File: js/110.balanced-binary-tree.js
/*
 * @lc app=leetcode id=110 lang=javascript
 *
 * [110] Balanced Binary Tree
 */
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isBalanced = function(root) {
	if (!root) return true;
	else if (!root.left && !root.right) return true;

	let left = getHeight(root.left);
	let right = getHeight(root.right);

	if (Math.abs(left - right) > 1) return false;
	else return isBalanced(root.left) && isBalanced(root.right);
};
// -1 is not balanced , otherwise is the
function getHeight(root) {
	if (!root) return 0;
	return Math.max(getHeight(root.left), getHeight(root.right)) + 1;
}

File: unsorted/java/110.balanced-binary-tree.java
/*
 * @lc app=leetcode id=110 lang=java
 *
 * [110] Balanced Binary Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode
 * left; TreeNode right; TreeNode(int x) { val = x; } }
 */
class Solution {
    public boolean isBalanced(TreeNode root) {

        return balancedTreeWithDepth(root) != -1;
    }

    // -1 no: [0..depth] yes
    int balancedTreeWithDepth(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int depthL = balancedTreeWithDepth(root.left);
        int depthR = balancedTreeWithDepth(root.right);
        if (depthL == -1 || depthR == -1)
            return -1;
        if (Math.abs(depthL - depthR) > 1)
            return -1;
        else
            return Math.max(depthL, depthR) + 1;
    }
}
// @lc code=end
